Refactor Join Expression Actions, add initial initial join order optimization

vdimir opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):

- Improvement


### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Added experimental join order optimization that can automatically reorder JOINs for better performance (controlled by `query_plan_optimize_join_order_limit` setting). Note that the join order optimization currently has limited statistics support and primarily relies on row count estimates from storage engines - more sophisticated statistics collection and cardinality estimation will be added in future releases. **If you encounter issues with JOIN queries after upgrading**, you can temporarily disable the new implementation by setting `SET query_plan_use_new_logical_join_step = 0` and report the issue for investigation. **Note about resolution of identifiers from USING clause**: Changed resolving of the coalesced column from `OUTER JOIN ... USING` clause to be more consistent: previously, when selecting both the USING column and qualified columns (`a, t1.a, t2.a`) in a OUTER JOIN, the USING column would incorrectly be resolved to `t1.a`, showing 0/NULL for rows from the right table with no left match. Now identifiers from USING clause are always resolved to the coalesced column, while qualified identifiers resolve to the non-coalesced columns, regardless of which other identifiers are present in the query.
For example:
```sql
SELECT a, t1.a, t2.a FROM (SELECT 1 as a WHERE 0) t1 FULL JOIN (SELECT 2 as a) t2 USING (a)
-- Before: a=0, t1.a=0, t2.a=2 (incorrect - 'a' resolved to t1.a)
-- After:  a=2, t1.a=0, t2.a=2 (correct - 'a' is coalesced)

SELECT t1.a FROM (SELECT 1 as a WHERE 0) t1 FULL JOIN (SELECT 2 as a) t2 USING (a)
-- Before: t1.a=2 (incorrect - qualified name resolved to coalesced column)
-- After:  t1.a=0 (correct - qualified name resolves to actual t1.a)
```
This behaviour is now aligned with other DBMSs like PostgreSQL.


### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

Continuing, close https://github.com/ClickHouse/ClickHouse/pull/77900

vdimir committed:
[wip] Refactor Join Expression Actions

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
try not force bypass all inputs in JoinStep dag

vdimir committed:
use echoOn in 01721_join_implicit_cast_long

vdimir committed:
f

vdimir committed:
update join using result in 01721_join_implicit_cast_long

vdimir committed:
do not join on const if equality is in join expression

vdimir committed:
upd 03357_join_pk_sharding

Sharding works with join_use_nulls, because pre join expression now have non nullable columns in output

vdimir committed:
fix join predicate pinning

vdimir committed:
add limit for join reodering graph size

vdimir committed:
upd 03275_auto_cluster_functions_with_parallel_replicas

Updated descrtiptions at expression step

vdimir committed:
apply read in order after join reordering

vdimir commented:
Tested on tphc with dummy statistics:
<details><summary>Details</summary>

```
    --max_memory_usage=35G \
    --allow_experimental_correlated_subqueries=1 \
    --param__internal_join_table_stat_hints='{
          "customer" : { "cardinality": 15000000,  "distinct_keys": { "c_custkey": 15000000 , "c_nationkey": 25 } }
        , "orders" :   { "cardinality": 22760819,  "distinct_keys": { "o_orderkey": 22760819 , "o_custkey": 8735689 , "o_orderdate": 365 } }
        , "lineitem" : { "cardinality": 600037902, "distinct_keys": { "l_orderkey": 150000000 , "l_partkey": 20000000 , "l_suppkey": 1000000 , "l_discount": 11 } }
        , "supplier" : { "cardinality": 1000000,   "distinct_keys": { "s_suppkey": 1000000 , "s_nationkey": 25 } }
        , "nation" :   { "cardinality": 25,        "distinct_keys": { "n_nationkey": 25 , "n_regionkey": 5 } }
        , "region" :   { "cardinality": 1,         "distinct_keys": { "r_regionkey": 1 , "r_name": 1 } }
        , "part" :     { "cardinality": 20000000,  "distinct_keys": { "p_partkey": 20000000 } }
        , "partsupp" : { "cardinality": 80000000, "distinct_keys": { "ps_partkey": 20000000 , "ps_suppkey": 1000000 } }
    }' \
    --query_plan_optimize_join_order=1  
```

</details> 

<img width="418" height="648" alt="image" src="https://github.com/user-attachments/assets/519e1b6f-f752-4020-b83e-7d25fa87fcf5" />

vdimir committed:
drop unused columns after join physical step

vdimir committed:
join order: simplify pinned and dependencies

vdimir committed:
f

vdimir committed:
f

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
fix outer join handling

vdimir committed:
fix 03274_join_algorithm_default/02236_explain_pipeline_join

Added estimation for LIMIT, for few rows non parallel join is chosen

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
swap any join support

vdimir committed:
fix tidy

vdimir committed:
fix 02151_hash_table_sizes_stats_joins

KochetovNicolai commented on src/Analyzer/Resolve/IdentifierResolver.cpp:
```diff
@@ -997,7 +1056,12 @@ IdentifierResolveResult IdentifierResolver::tryResolveIdentifierFromJoin(const I
             throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected node type {}, expected function node", node->getNodeType());
 
-        const auto & function_argument_nodes = node->as<FunctionNode &>().getArguments().getNodes();
+        const auto & function_node = node->as<FunctionNode &>();
```
TODO: maybe check if `check_nested_column_not_in_using` is needed at all.
I think the issue is only with the FULL JOIN. But if `firstTruthy` correctly deduces the type, subcolumns/nested should work (or throw an exception).

Also maybe now we can apply an optimization, like `getSubcolumn(firstTruthy(*), 'name') -> firstTruthy(getSubcolumn(*, 'name))`

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -310,12 +309,7 @@ std::optional<JoinTableSide> QueryAnalyzer::getColumnSideFromJoinTree(const Quer
-                    resolved_identifier->formatASTForErrorMessage(),
-                    join_node.formatASTForErrorMessage());
-            }
+                return {};
```
I can hardly imagine in which case we could end up here before.
Now, it's firstTruthy?
Does it make sense still throw an exception in other cases?

KochetovNicolai commented on src/Analyzer/Resolve/QueryAnalyzer.cpp:
```diff
@@ -2166,18 +2168,11 @@ QueryAnalyzer::QueryTreeNodesWithNames QueryAnalyzer::resolveUnqualifiedMatcher(
-                    correctColumnExpressionType(matched_column_node->as<ColumnNode &>(), scope.context);
-                    if (!matched_column_node->isEqual(*join_using_column_nodes.at(0)))
-                        scope.join_columns_with_changed_types[matched_column_node] = join_using_column_nodes.at(0);
+                    // if (!matched_column_node->isEqual(*join_using_column_nodes.at(0)))
```
Remove?

KochetovNicolai commented on src/Common/safe_cast.h:
```diff
@@ -10,13 +15,25 @@ namespace ErrorCodes
-    if (from > max)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Overflow ({} > {})", from, max);
-    return static_cast<To>(from);
+    if constexpr (std::is_signed_v<From> == std::is_signed_v<To> &&
```
But actually this is ok.

KochetovNicolai commented on src/Functions/firstTruthy.cpp:
```diff
@@ -0,0 +1,161 @@
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
+    {
+        if (arguments.empty())
```
This is impossible

KochetovNicolai commented on src/Functions/firstTruthy.cpp:
```diff
@@ -0,0 +1,161 @@
+            return result_type->createColumnConstWithDefaultValue(input_rows_count);
+
+        if (arguments.size() == 1)
+            return castColumn(arguments[0], result_type);
```
Cast is not needed

KochetovNicolai commented on src/Functions/firstTruthy.cpp:
```diff
@@ -0,0 +1,161 @@
+                /// - for numeric types, the default is 0
+                /// - for strings, the default is ''
+                /// - for arrays, the default is []
+                if (!arguments[arg_idx].column->isNullAt(row) && !arguments[arg_idx].column->isDefaultAt(row))
```
Null should be the default for a nullable column; probably, the first check is not needed.

Also, 0 is not a default for Nullable column, but it's not "truthy" in my opinion.
Maybe rename this function to `firstNotDefault`?

KochetovNicolai commented on src/Functions/firstTruthy.cpp:
```diff
@@ -0,0 +1,161 @@
+            if (!found)
+            {
+                const auto & last_column = casted_columns.back();
+                result_col->insertFrom(*last_column, row);
```
Funny thing that for `(0 : Int, Null : Nullable)` it returns `Null`. And this is correct.
But for `(Null : Nullable, 0: Int)` it returns `0`, which is not correct in terms of CROSS JOIN (it could be non-matched Null key from the left table).

It seems to be more correct to `->insertDefault()` here?

KochetovNicolai commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -3577,18 +3602,69 @@ static MutableColumnPtr deserializeConstant(
+    return node_to_id;
+}
+
+std::unordered_map<size_t, const ActionsDAG::Node *> ActionsDAG::getIdToNodeMap() const
```
If to think about it, we can just return `std::vector<const ActionsDAG::Node *>`

KochetovNicolai commented on src/Interpreters/ActionsDAG.cpp:
```diff
@@ -3838,4 +3914,12 @@ ActionsDAG ActionsDAG::deserialize(ReadBuffer & in, DeserializedSetsRegistry & r
     return dag;
 }
 
+bool ActionsDAG::containsNode(const Node * node)
```
I prefer to avoid such an O(n) lookups.
It's only used once now, so probably we can do it there

KochetovNicolai commented on src/Interpreters/ActionsDAG.h:
```diff
@@ -459,6 +472,10 @@ class ActionsDAG
     UInt64 getHash() const;
     void updateHash(SipHash & hash_state) const;
 
+    static std::optional<ActionsForFilterPushDown> createActionsForConjunction(NodeRawConstPtrs conjunction, const ColumnsWithTypeAndName & all_inputs);
```
It's not private anymore, so let's add a comment

KochetovNicolai commented on src/Interpreters/JoinExpressionActions.cpp:
```diff
@@ -0,0 +1,514 @@
+    size_t pos = bitset.find_first();
+    size_t maxPos = pos;
+
+    while (pos != Base::npos)
+    {
+        maxPos = pos;
+        pos = bitset.find_next(pos);
+    }
```
I am a bit confused about this method.
Why do we do an iteration here? Should it be called `find_last`?
Why do we return int, not size_t or optional<size_t>?

KochetovNicolai committed:
Keep only one map.

KochetovNicolai commented on src/Interpreters/JoinExpressionActions.cpp:
```diff
@@ -0,0 +1,514 @@
+    return static_cast<int>(maxPos);
+}
+
+size_t BitSet::hashImpl() const
```
I don't really like using BitSet as a key of unordered_map.
Now, it is used in two places: 
* dynamic programming (maybe we can use size_t instead of Set; it can work only for a few tables anyway
* to calculate names; I am sure it can be done without it

So, maybe we can remove it.

KochetovNicolai commented on src/Interpreters/JoinExpressionActions.cpp:
```diff
@@ -0,0 +1,514 @@
+    Data::NodeToSourceMapping expression_sources;
+
+    const auto & input_nodes = actions_dag_.getInputs();
+    if (input_nodes.size() != left_header.columns() + right_header.columns())
```
Todo: Here, we can (probably) enforce an even stricter invariant that inputs are ordered in the same way.
(or just use size_t to split left and right inputs)

KochetovNicolai commented on src/Planner/CollectTableExpressionData.cpp:
```diff
@@ -70,9 +70,9 @@ class CollectSourceColumnsVisitor : public InDepthQueryTreeVisitorWithContext<Co
-            if (columns_from_subtrees.size() != 2)
-                throw Exception(ErrorCodes::LOGICAL_ERROR,
-                    "Expected two columns in JOIN using expression for column {}", column_node->dumpTree());
+            // if (columns_from_subtrees.size() != 2)
```
Remove?

KochetovNicolai commented on src/Planner/PlannerJoins.cpp:
```diff
@@ -1059,7 +1059,13 @@ std::shared_ptr<DirectKeyValueJoin> tryDirectJoin(const std::shared_ptr<TableJoi
 
-    for (const auto & right_table_expression_column : *right_table_expression_header)
+    Block right_table_expression_header_ = *right_table_expression_header;
+    // if (table_join->joinUseNulls() && isLeftOrFull(table_join->kind()))
```
Remove?

KochetovNicolai committed:
Merge branch 'master' into vdimir/small_refactoring

KochetovNicolai committed:
Merge remote-tracking branch 'origin/master' into vdimir/small_refactoring

KochetovNicolai committed:
Fixing build

KochetovNicolai commented on src/Processors/QueryPlan/JoinStepLogical.h:
```diff
@@ -44,13 +53,25 @@ class JoinStepLogical final : public IQueryPlanStep
         JoinExpressionActions join_expression_actions_,
-        Names required_output_columns_,
+        const NameSet & required_output_columns_,
+        const std::unordered_map<String, const ActionsDAG::Node *> & changed_types,
```
Let's add a comment on what `changed_types` is.

Now, it's only needed for `using` columns, to cast those into supertype.
E.g. for `JOIN ... using (x)` the `join_expression_actions` dag may contain chain like `INPUT x -> CAST(x, supertype) -> ALIAS x`; then `changed_types` will contain `"x" -> const Node * ALIAS x`

vdimir committed:
Fix join step with using reordering for distributed plan

vdimir committed:
fix tests

vdimir committed:
fixes

vdimir committed:
[tmp] runn fuzzers and stress early

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
upd SettingsChangesHistory.cpp

vdimir committed:
Replace logical error to not found due to bug in par replicas

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp:
```diff
@@ -44,18 +57,138 @@ namespace Setting
+
+void remapColumnStats(std::unordered_map<String, ColumnStats> & mapped, const ActionsDAG & actions)
+{
+    std::unordered_map<String, ColumnStats> original = std::move(mapped);
+    mapped = {};
+    for (const auto & [name, value] : original)
+    {
+        for (const auto & remapped : backTrackColumnsInDag(name, actions))
+            mapped[remapped] = value;
+    }
```
Also, I don't like how `firstTruthy` is handled.
If we have statistics for both args, maybe we should get `max` from both (or even something smarter). If some size is unavailable, we probably need to ignore statistics. Because right now it's FULL JOIN case.

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp:
```diff
@@ -159,247 +322,731 @@ bool optimizeJoinLegacy(QueryPlan::Node & node, QueryPlan::Nodes &, const QueryP
+
+    std::vector<JoinActionRef> join_edges;
+
+    std::unordered_map<size_t, std::pair<BitSet, JoinKind>> join_kinds;
+    std::unordered_map<size_t, ActionsDAG::NodeRawConstPtrs> type_changes;
```
Do we use input number as a key? If so, maybe we can put `join_kinds, type_changes, relation_stats, inputs` into one  `QueryGraphBuilder::Node` structure.

KochetovNicolai commented on src/Processors/QueryPlan/JoinStepLogical.cpp:
```diff
@@ -40,85 +68,148 @@ namespace Setting
+    }
+}
+
+std::unordered_set<JoinTableSide> JoinStepLogical::typeChangingSides() const
```
Can we please replace `std::unordered_set<JoinTableSide>` to enum with 4 values?

KochetovNicolai commented on src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp:
```diff
@@ -159,247 +322,731 @@ bool optimizeJoinLegacy(QueryPlan::Node & node, QueryPlan::Nodes &, const QueryP
-    else if (optimization_settings.join_swap_table.value())
+
+    for (const auto & out : global_actions_dag->getOutputs())
+        usage_level_map[out] = sequence.size();
```
Can we override the previous value from expression/residual filter? E.g. in `select l.x from l join r on l.x = r.x and l.x`?

vdimir replied:
Yes, I think so, it's intended. We need to take latest stage where node is used.

vdimir committed:
fix some teste for old analyzer

vdimir committed:
disable join reordering with par replicas

vdimir committed:
Rename firstTruthy -> firstNotDefault

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

KochetovNicolai reviewed (approved):
Overall looks good.

I did not review `joinOrder.cpp`.

vdimir committed:
join reorder settings, fixes, tests

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
fixing review comments

vdimir committed:
revert ReadFromSystemNumbersStep

vdimir committed:
better usage_level_map in optimizeJoin

vdimir commented on src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp:
```diff
@@ -159,247 +322,731 @@ bool optimizeJoinLegacy(QueryPlan::Node & node, QueryPlan::Nodes &, const QueryP
+            for (const auto * child : node->children)
+                stack.push(std::make_pair(child, level));
+            auto & map_level = usage_level_map[node];
+            map_level = std::max(map_level, level);
```
We can traverse nodes from higher levels to lower and skip vising them twice: https://github.com/ClickHouse/ClickHouse/pull/80848/commits/7e8483083c29415330eab9626cb738702e35fd98

vdimir committed:
disable par replicas in tests swaping joined tables

vdimir committed:
join graph: findFirstSet -> getSingleBit

vdimir committed:
remove some unused code

vdimir committed:
fix stylecheck

vdimir committed:
fix clang tidy, upd firstNonDefault.cpp

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
update tests

The 03006_join_on_inequal_expression_fast_1 used to return different results for hash join due to different order of clauses in hash join.

vdimir committed:
revert changes in ci

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
parse table aliases in dummy join stats

vdimir committed:
add tests/performance/join_table_order.xml

vdimir committed:
small fixes

vdimir committed:
upd perf tests

join_table_order does not work on previous version due to unknonw setting,
need add it later

vdimir committed:
no-asan 01062_pm_all_join_with_block_continuation

vdimir commented:
https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80848&sha=3051a5d4b71b284d4eda27ca032811c6210b9ca4&name_0=PR&name_1=Stateless+tests+%28amd_ubsan%2C+parallel%29

`02177_issue_31009` looks suspicious but actually not related, already flaky: https://github.com/ClickHouse/ClickHouse/issues/86335

vdimir committed:
minor changes

vdimir committed:
Merge branch 'master' into vdimir/small_refactoring

vdimir committed:
upd SettingsChangesHistory.cpp

vdimir committed:
fix typo

vdimir commented:
### The following tests have been updated with their corresponding reasons:

- `00057_join_aliases` `00779_all_right_join_max_block_size` `01010_pmj_on_disk`
Setting query_plan_join_swap_table started to work for USING and `true` can break the test

- `00848_join_use_nulls_segfault` `01115_join_with_dictionary` `01323_redundant_functions_in_order_by` `01476_right_full_join_switch` `01477_lc_in_merge_join_left_key` `02374_analyzer_join_using` `02722_matcher_join_use_nulls` `02771_semi_join_use_nulls` `02955_analyzer_using_functional_args` `02955_analyzer_using_functional_args_oldanalyzer`
Identifier from USING vs with qualified identifiers

- `01062_pm_all_join_with_block_continuation`
Does not pass flaky check: https://github.com/ClickHouse/ClickHouse/pull/85132

- `01101_literal_column_clash` `01478_not_equi` `03362_join_where_false_76670`
Constant JOIN ON condition (after pushing down) evaluated for LEFT JOIN, previously it was `INVALID_JOIN_ON_EXPRESSION`

- `01332_join_type_syntax_position`
Changed error code

- `01655_plan_optimizations`
Changed EXPLAIN output

- `01872_functions_to_subcolumns_analyzer`
Updated `EXPLAIN QUERY TREE` with `firstNonDefault`

- `02236_explain_pipeline_join` `03274_join_algorithm_default`
Changed EXPLAIN PIPELINE, increased limit to keep ConcurrentHashJoin and not choose HashJoin

- `02514_analyzer_drop_join_on` `02675_predicate_push_down_filled_join_fix` `02835_join_step_explain` `02967_parallel_replicas_joins_and_analyzer` `03036_join_filter_push_down_equivalent_sets` `03130_convert_outer_join_to_inner_join` `03152_join_filter_push_down_equivalent_columns` `03275_auto_cluster_functions_with_parallel_replicas` `03355_join_to_in_optimization` `03357_join_pk_sharding` `03405_merge_filter_into_join` `03545_analyzer_correlated_subqueries_use_equivalence_classes` `03545_analyzer_correlated_subqueries_use_equivalence_classes_2` `03545_analyzer_correlated_subquery_exists_asterisk`
Changed expression steps in EXPLAIN PLAN

- `01721_join_implicit_cast_long` `01721_join_implicit_cast_long_oldanalyzer`
Updated USING: https://github.com/ClickHouse/ClickHouse/pull/80848/commits/4584fed0a420494ed6debc1a0b1ac7a99f1ea8e7 + https://github.com/ClickHouse/ClickHouse/pull/80848/commits/487d6b28755fc08f91cd49f04ee56699767c4c7e

- `02151_hash_table_sizes_stats_joins` `03279_join_choose_build_table`
Updated test to show query id in case of failure

- `02177_issue_31009`
SET query_plan_join_swap_table = 0, but probably the test is flaky in master

- `02273_full_sort_join` `02276_full_sort_join_unsupported`
USING + join_use_nulls was not working for full_sorting_merge

- `02381_join_dup_columns_in_plan`
*Disabled reordering for test, but seems now it works and we can re-enable it

- `02479_analyzer_join_with_constants`
ASOF JOIN on constant works instead of `INVALID_JOIN_ON_EXPRESSION`

- `03006_join_on_inequal_expression_fast_1`
SEMI JOIN with several alternative conditions, order of clauses was inverted, so lead to different row chosen https://fiddle.clickhouse.com/88f83396-0af0-40ad-8bd9-9f8cadaa010c

- `03279_pr_3_way_joins_inner_first` `03279_pr_3_way_joins_left_first`
*It's USING vs qualified identifiers, but also test reference was huge with some unexpected values were canonized (probably it's a bug with Parallel replicas, needs separate investigation, previously discussed with @devcrafter), for now replaced output with hash  

- `03290_alter_mv_dependency_check`
Not related, just noticed that *.gen.reference was committed instead of .reference
   
- `03356_threshold_for_parallel_hash` `03593_any_join_swap_tables`
Join optimization (and table size estimation) disabled with parallel replicas

- `03402_join_using_alias`
*Query started to execute, JOIN USING alias with distributed, also related to https://github.com/ClickHouse/ClickHouse/pull/83921, need to revisit that PR

- `03522_function_first_non_default` `03523_function_first_non_default_number_args`
Added test for new function

- `03524_join_multiple_using`
*Committed reference, but not a test, need to fix
